File: backend/app/routers/stats.py
from datetime import date, timedelta
from decimal import Decimal

from fastapi import APIRouter, Depends
from sqlalchemy import func, select
from sqlalchemy.orm import Session

from app.database import get_db
from app.models import PC, Part, PartCondition, PCStatus, TransferLog
from app.schemas import StatsOut, TransferLogOut

router = APIRouter(tags=["dashboard"])


@router.get("/stats", response_model=StatsOut)
def get_stats(db: Session = Depends(get_db)):
    today = date.today()
    cutoff = today + timedelta(days=30)
    total_pcs = db.scalar(select(func.count(PC.id))) or 0
    active_pcs = (
        db.scalar(select(func.count(PC.id)).where(PC.status == PCStatus.active)) or 0
    )
    total_parts = db.scalar(select(func.count(Part.id))) or 0
    inventory_count = (
        db.scalar(
            select(func.count(Part.id))
            .where(Part.pc_id.is_(None))
            .where(Part.employee_id.is_(None))
        )
        or 0
    )
    total_value = db.scalar(select(func.sum(Part.purchase_price))) or Decimal("0")
    expiring = (
        db.scalar(
            select(func.count(Part.id))
            .where(Part.warranty_expiry.is_not(None))
            .where(Part.warranty_expiry >= today)
            .where(Part.warranty_expiry <= cutoff)
        )
        or 0
    )
    faulty = (
        db.scalar(
            select(func.count(Part.id)).where(
                Part.condition.in_([PartCondition.faulty, PartCondition.rma])
            )
        )
        or 0
    )
    return StatsOut(
        total_pcs=total_pcs,
        active_pcs=active_pcs,
        total_parts=total_parts,
        inventory_count=inventory_count,
        total_value=total_value,
        expiring_warranties=expiring,
        faulty_parts=faulty,
    )


@router.get("/transfers/recent", response_model=list[TransferLogOut])
def recent_transfers(db: Session = Depends(get_db), limit: int = 10):
    logs = (
        db.execute(
            select(TransferLog)
            .order_by(TransferLog.moved_at.desc())
            .limit(min(limit, 50))
        )
        .scalars()
        .all()
    )
    pc_ids = {pid for log in logs for pid in (log.from_pc_id, log.to_pc_id) if pid}
    names: dict[str, str] = {}
    if pc_ids:
        rows = db.execute(select(PC.id, PC.name).where(PC.id.in_(pc_ids))).all()
        names = {row.id: row.name for row in rows}
    out = []
    for log in logs:
        item = TransferLogOut.model_validate(log)
        item.from_pc_name = names.get(log.from_pc_id) if log.from_pc_id else None
        item.to_pc_name = names.get(log.to_pc_id) if log.to_pc_id else None
        if log.part:
            item.part_label = f"{log.part.brand} {log.part.model}"
        out.append(item)
    return out

File: backend/routes/stats.js
import { Router } from 'express';
import { PC, Part, TransferLog } from '../config/models.js';

const router = Router();

router.get('/stats', async (req, res, next) => {
  try {
    const today = new Date();
    const cutoff = new Date(today.getTime() + 30 * 86400000);
    const allParts = await Part.find({}).lean();
    const allPcs = await PC.find({}).lean();
    const totalPcs = allPcs.length;
    const activePcs = allPcs.filter(p => p.status === 'active').length;
    const totalParts = allParts.length;
    const inventoryCount = allParts.filter(p => !p.pc_id && !p.employee_id).length;
    const totalValue = allParts.reduce((s, p) => s + (p.purchase_price || 0), 0);
    const expiring = allParts.filter(p => p.warranty_expiry && new Date(p.warranty_expiry) >= today && new Date(p.warranty_expiry) <= cutoff).length;
    const faulty = allParts.filter(p => p.condition === 'faulty' || p.condition === 'rma').length;
    res.json({ total_pcs: totalPcs, active_pcs: activePcs, total_parts: totalParts, inventory_count: inventoryCount, total_value: totalValue, expiring_warranties: expiring, faulty_parts: faulty });
  } catch (err) { next(err); }
});

router.get('/transfers/recent', async (req, res, next) => {
  try {
    const limit = Math.min(parseInt(req.query.limit) || 10, 50);
    let logs = await TransferLog.find({}).lean();
    logs.sort((a, b) => new Date(b.moved_at) - new Date(a.moved_at));
    logs = logs.slice(0, limit);
    const pcIds = [...new Set(logs.flatMap(l => [l.from_pc_id, l.to_pc_id].filter(Boolean)))];
    const names = {};
    for (const id of pcIds) { const p = await PC.findById(id).lean(); if (p) names[id] = p.name; }
    const result = [];
    for (const log of logs) {
      const item = { ...log, id: log._id, from_pc_name: log.from_pc_id ? names[log.from_pc_id] || null : null, to_pc_name: log.to_pc_id ? names[log.to_pc_id] || null : null };
      if (log.part_id) { const part = await Part.findById(log.part_id).lean(); item.part_label = part ? `${part.brand} ${part.model}` : null; }
      result.push(item);
    }
    res.json(result);
  } catch (err) { next(err); }
});

export default router;
